### ISSUE TITLE
[Bug]: A5 BiShengHIR fails on memref.extract_strided_metadata when tl.sum(..., axis=1) follows masked 2D load with runtime base offset

### ISSUE THREAD

**Issue by zheliuyu** [triton-lang/triton-ascend#2239]
### Summary

On **Ascend A5** (Ascend950PR, `Ascend950PR_9579` compile target), a minimal Triton kernel **fails to compile** during the BiShengHIR pipeline when all of the following hold:

1. **Runtime pointer base offset**: `task_base = program_id(0) * hidden_size` added to 2D tile indices (typical multi-task / batch loop pattern).
2. **Padded 2D masked load**: `BLOCK_H = next_power_of_2(hpc) > hpc` with `channels_per_group = 1` (e.g. `hpc = 3`, `BLOCK_H = 4`, `BLOCK_CH = 1`).
3. **Reduction with `axis=1`**: `tl.sum(tl.sum(x, axis=1), axis=0)` (equivalently `tl.sum(x, axis=1)` in a 1-row tile).

The **same** load + **axis-free** `tl.sum(x)` compiles and runs. The **same** `axis=1` reduction with a **static** base pointer (no `task_base`) also compiles on A5. The failure is a **compile-time IR type inference error**, not a runtime numeric mismatch or UB overflow.

This pattern appears in downstream kernels (e.g. fused GroupNorm mean/variance reduction). On **A3**, the identical Triton source is reported to compile via the A2/A3 backend path; only the **A5** path (`bishengir-compile-a5`, dynamic CV pipeline) reproduces the failure.

### Minimal complete example

Save as `repro_gn_axis1_dynamic_offset.py` and run on an idle A5 NPU:

```bash
python repro_gn_axis1_dynamic_offset.py
```

```python
#!/usr/bin/env python3
"""Minimal A5 compile repro."""

import sys

import torch
import triton
import triton.language as tl

HIDDEN = 3
BLOCK_H = 4
BLOCK_CH = 1
CPG = 1


def ir_or_ok(label: str, fn) -> None:
    try:
        fn()
        torch.npu.synchronize()
        print(f"{label}: OK")
    except Exception as e:
        text = str(e)
        if "memref.extract_strided_metadata" in text:
            for ln in text.splitlines():
                if "memref.extract_strided_metadata" in ln:
                    print(f"{label}: FAIL (IR)")
                    print(" ", " ".join(ln.split())[:240])
                    return
        print(f"{label}: FAIL ({type(e).__name__})")
        print(" ", text.splitlines()[-1][:200] if text.splitlines() else text)


@triton.jit
def control_sum_all(
    Out,
    Inp,
    hidden_size: tl.constexpr,
    channels_per_group: tl.constexpr,
    BLOCK_H: tl.constexpr,
    BLOCK_CH: tl.constexpr,
):
    hpc: tl.constexpr = hidden_size // channels_per_group
    pid = tl.program_id(0)
    ch = tl.arange(0, BLOCK_CH)
    h = tl.arange(0, BLOCK_H)
    mask = (ch < channels_per_group)[:, None] & (h < hpc)[None, :]
    task_base = pid * hidden_size
    ptrs = task_base + ch[:, None] * hpc + h[None, :]
    x = tl.load(Inp + ptrs, mask=mask, other=0.0).to(tl.float32)
    tl.store(Out + pid, tl.sum(x))


@triton.jit
def fail_sum_axis1(
    Out,
    Inp,
    hidden_size: tl.constexpr,
    channels_per_group: tl.constexpr,
    BLOCK_H: tl.constexpr,
    BLOCK_CH: tl.constexpr,
):
    hpc: tl.constexpr = hidden_size // channels_per_group
    pid = tl.program_id(0)
    ch = tl.arange(0, BLOCK_CH)
    h = tl.arange(0, BLOCK_H)
    mask = (ch < channels_per_group)[:, None] & (h < hpc)[None, :]
    task_base = pid * hidden_size
    ptrs = task_base + ch[:, None] * hpc + h[None, :]
    x = tl.load(Inp + ptrs, mask=mask, other=0.0).to(tl.float32)
    tl.store(Out + pid, tl.sum(tl.sum(x, axis=1), axis=0))


def main() -> int:
    torch.npu.set_device(0)
    print("device:", torch.npu.get_device_name(0))
    print(f"constexpr hidden={HIDDEN} BLOCK_H={BLOCK_H} BLOCK_CH={BLOCK_CH} cpg={CPG}")
    x = torch.randn(HIDDEN, device="npu", dtype=torch.float32)
    out = torch.empty(1, device="npu", dtype=torch.float32)
    grid = (1,)
    ir_or_ok("control_sum_all", lambda: control_sum_all[grid](out, x, HIDDEN, CPG, BLOCK_H, BLOCK_CH))
    ir_or_ok("fail_sum_axis1", lambda: fail_sum_axis1[grid](out, x, HIDDEN, CPG, BLOCK_H, BLOCK_CH))
    return 0


if __name__ == "__main__":
    sys.exit(main())
```

### Expected vs observed

**Expected**

- Both kernels compile on A5; `fail_sum_axis1` stores the sum of the three loaded elements (same scalar as `control_sum_all` for this shape).

**Observed (A5, Ascend950PR_9579)**

```
device: Ascend950PR_9579
constexpr hidden=3 BLOCK_H=4 BLOCK_CH=1 cpg=1
control_sum_all: OK
fail_sum_axis1: FAIL (IR)
```

Compiler diagnostics (abbreviated; full log includes IR dump after `InsertLoadStoreForMixCV`):

```
loc(".../repro_gn_axis1_dynamic_offset.py":88:16): error: 'memref.extract_strided_metadata' op failed to infer returned types
loc(".../repro_gn_axis1_dynamic_offset.py":88:16): error: 'memref.extract_strided_metadata' op inferred type(s) 'memref<f32>', 'index', 'index', 'index' are incompatible with return type(s) of operation 'memref<f32>', 'index', 'index', 'index', 'index', 'index'
loc(".../repro_gn_axis1_dynamic_offset.py":73:0): error: Failed to run BiShengHIR pipeline
[ERROR] Failed to run BiShengIR pipeline
```

Representative failing IR site (after masked tile is collapsed to length-3 1D memref):

```
%12 = "memref.subview"(%10) ... : (...) -> memref<3xf32, strided<[1], offset: ?>>
%14:6 = "memref.extract_strided_metadata"(%12) : (memref<3xf32, strided<[1], offset: ?>>) -> (memref<f32>, index, index, index, index, index)
```

The op is inferred as **4** result types (1D rank) but the operation expects **6** (2D rank).

Invoked backend (from Triton error footer):

```
bishengir-compile-a5 ... --target=Ascend950PR_9579 --enable-auto-multi-buffer=True ... --enable-triton-kernel-compile=true ...
```

### Bisection matrix (same constexpr, A5)

| Variant | `hidden=3`, `BLOCK_H=4`, `BLOCK_CH=1` |
|--------|----------------------------------------|
| Dynamic `task_base` + `tl.sum(x)` | OK |
| Dynamic `task_base` + `tl.sum(..., axis=1)` | **Compile fail (IR above)** |
| Static pointer (no `task_base`) + `tl.sum(..., axis=1)` | OK |
| Dynamic + `axis=1`, `hidden=4` (`BLOCK_H == hpc`) | OK |
| Dynamic + `axis=1`, `hidden=2`, `BLOCK_H=4` (`BLOCK_H > hpc`) | **Compile fail** |

Additional local scripts (optional attachment): `min_repro_gn_sum.py`, `min_repro_group_norm_dyn.py` in our downstream repo experiments folder.

### Downstream impact

In [Liger-Kernel](https://github.com/linkedin/Liger-Kernel) Ascend GroupNorm, shape `(N,C,H)=(1,1,1,3)` hits the single-task forward kernel with `cpg=1`, `hpc=3`, `BLOCK_H=4`. Pytest `test_liger_group_norm[..., 1-1-1-3]` failed on A5 until mean/variance used axis-free `tl.sum(X_vals)` / `tl.sum(X_vals * X_vals)` (mathematically equivalent for `cpg=1`). We prefer a compiler fix rather than permanently avoiding `axis=1` in generic 2D tiles.

### Workaround (application side)

For `channels_per_group == 1`, use `tl.sum(x)` instead of `tl.sum(tl.sum(x, axis=1), axis=0)`. This is not acceptable as the only general fix when `cpg > 1` layouts still use 2D tiles and axis reductions.

### Environment details

```
Triton: 3.2.0
triton_ascend: 3.2.2 (pip package name triton_ascend)
PyTorch: 2.9.0+cpu
torch_npu: 2.9.0.post4
NPU: Ascend950PR (reported device name: Ascend950PR_9579)
OS: Linux (openEuler 24.03 SP3, kernel 6.6.x)
Python: 3.11
```

Compile path: 

A5 — bishengir-compile-a5 --target=Ascend950PR_9579 (with enable-auto-multi-buffer, triton-kernel-compile, hfusion, etc. as in default triton-ascend A5 pipeline)

Cross-SOC: 

Same Triton kernel source compiles on Ascend A3 (A2/A3 pipeline); failure observed only on A5 pipeline above.

Repro command:
```
  python repro_gn_axis1_dynamic_offset.py
  # first JIT compile ~15–30s; control kernel OK, failing kernel throws during compile (no kernel launch)
```

**Comment by hongziqi** [comment]
Thanks for your feedback. We will arrange relevant engineers to investigate this issue later.

**Comment by zhudada0120** [comment]
Hi @zheliuyu, Thanks for your feedback. We have reproduced the issue with the minimal reproduction case you provided. Our analysis shows that the generated adapter IR is identical on A3 and A5 hardware. This issue is preliminarily ruled out as a problem within the triton-ascend compilation flow. We recommend submitting an issue to the NPUIR community for further tracking and troubleshooting.
NPUIR Community Link: [https://gitcode.com/Ascend/AscendNPU-IR](https://link.wtturl.cn/?target=https%3A%2F%2Fgitcode.com%2FAscend%2FAscendNPU-IR&scene=im&aid=582478&lang=zh)

**Comment by zheliuyu** [comment]
@zhudada0120 Thank you for pinpointing this issue. I've reported it to the AscendNPU-IR community: https://gitcode.com/Ascend/AscendNPU-IR/issues/590. Waiting for their response.

**Comment by zheliuyu** [comment]
Let's keep this issue open for now, since the root cause hasn't been pinpointed yet.
